reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java

```diff
@@ -0,0 +1,299 @@
+						actual.onNext(result);
+						innerComplete();
+						return;
+					}
+
+					inner.set(new WeakScalarSubscription<>(result, inner));
+					return;
+				}
+
+				p.subscribe(inner);
+			}
+			catch (Throwable e) {
+				Context ctx = actual.currentContext();
+				Operators.onDiscard(t, ctx);
+
+				Throwable e_ = Operators.onNextError(t, e, ctx, upstream);
+				if (e_ == null) {
+					innerComplete();
+					return;
+				}
+				STATE.lazySet(this, State.TERMINATED);
+				actual.onError(Operators.onOperatorError(upstream, e, t, ctx));
+				return;
+			}
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
+					case CANCELED:
+					case TERMINATED:
+						Operators.onErrorDropped(t, currentContext());
+						return;
+					default:
+						if (!STATE.compareAndSet(this, previousState, State.TERMINATED)) {
+							continue;
+						}
+						inner.cancel();
+						synchronized (this) {
+							actual.onError(t);
+						}
+						return;
+				}
+			}
+		}
+
+		@Override
+		public void onComplete() {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
+					case INITIAL:
+					case REQUESTED:
+						if (!STATE.compareAndSet(this, previousState, State.TERMINATED)) {
+							continue;
+						}
+						actual.onComplete();
+						return;
+					case ACTIVE:
+						if (!STATE.compareAndSet(this, previousState, State.LAST_ACTIVE)) {
+							continue;
+						}
+						return;
+					default:
+						return;
+				}
+			}
+		}
+
+		@Override
+		public synchronized void innerNext(R value) {
+			switch (state) {
+				case ACTIVE:
+				case LAST_ACTIVE:
+					actual.onNext(value);
+					break;
+				default:
+					Operators.onDiscard(value, currentContext());
+					break;
+			}
+		}
+
+		@Override
+		public void innerComplete() {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
+					case ACTIVE:
+						if (!STATE.compareAndSet(this, previousState, State.REQUESTED)) {
+							continue;
+						}
+						upstream.request(1);
+						return;
+					case LAST_ACTIVE:
+						if (!STATE.compareAndSet(this, previousState, State.TERMINATED)) {
+							continue;
+						}
+						actual.onComplete();
+						return;
+					default:
+						return;
+				}
+			}
+		}
+
+		@Override
+		public void innerError(Throwable e) {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
+					case ACTIVE:
+					case LAST_ACTIVE:
+						if (!STATE.compareAndSet(this, previousState, State.TERMINATED)) {
+							continue;
+						}
+						upstream.cancel();
+						inner.cancel();
+						actual.onError(e);
+						return;
+					default:
+						Operators.onErrorDropped(e, currentContext());
+						return;
+				}
+			}
+		}
+
+		@Override
+		public void request(long n) {
+			if (STATE.compareAndSet(this, State.INITIAL, State.REQUESTED)) {
```

Comment:
here I think it is worth it detailing the behavior in a comment:
 - the fact that obviously the only time a downstream request directly results in an `upstream` request is the first time
 - detail of what happens when `request(10)` is done but the inner only serves eg. 3 elements (3 inner next will be tracked as produced by the `ConcatMapInner`, then inner onComplete will update the `requested` tracker, then `innerComplete` will transition to REQUESTED state and request one more source, swap the subscription and request the 7 remaining demand)

Reply:
(this is absolutely not necessary, but after reviewing that was the one spot I had to double back over)